UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AuthTokenProvider.cs

```diff
@@ -1,30 +1,86 @@
 ﻿using Azure.Core;
 using Azure.Identity;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using System;
 using System.Diagnostics.CodeAnalysis;
 using System.Threading.Tasks;
 using UKHO.ExchangeSetService.Common.Configuration;
+using UKHO.ExchangeSetService.Common.Logging;
+using UKHO.ExchangeSetService.Common.Models.Request;
 
 namespace UKHO.ExchangeSetService.Common.Helpers
 {
     [ExcludeFromCodeCoverage] ////Excluded from code coverage as it has ADD interaction
-    public class AuthTokenProvider : IAuthTokenProvider
+    public class AuthTokenProvider
     {
         private readonly IOptions<EssManagedIdentityConfiguration> essManagedIdentityConfiguration;
+        private readonly ILogger<AuthTokenProvider> logger;
+        private static readonly Object _lock = new Object();
+        private readonly IDistributedCache _cache;
 
-        public AuthTokenProvider(IOptions<EssManagedIdentityConfiguration> essManagedIdentityConfiguration)
+        public AuthTokenProvider(IOptions<EssManagedIdentityConfiguration> essManagedIdentityConfiguration, IDistributedCache _cache, ILogger<AuthTokenProvider> logger)
         {
             this.essManagedIdentityConfiguration = essManagedIdentityConfiguration;
+            this._cache = _cache;
+            this.logger = logger;
         }
 
+
         public async Task<string> GetManagedIdentityAuthAsync(string resource)
+        {
+            var accessToken = GetFromCache(resource);
+
+            if (accessToken != null && accessToken.AccessToken != null && accessToken.ExpiresIn > DateTime.UtcNow)
+            {
+                return accessToken.AccessToken;
+            }
+
+            var newAccessToken = await GetNewAuthToken(resource);
+            AddToCache(resource, newAccessToken);
+
+            return newAccessToken.AccessToken;
+        }
+
+        private async Task<AccessTokenItem> GetNewAuthToken(string resource)
         {
             var tokenCredential = new DefaultAzureCredential(new DefaultAzureCredentialOptions { ManagedIdentityClientId = essManagedIdentityConfiguration.Value.ClientId });
             var accessToken = await tokenCredential.GetTokenAsync(
                 new TokenRequestContext(scopes: new string[] { resource + "/.default" }) { }
             );
 
-            return accessToken.Token;
+            return new AccessTokenItem
+            {
+                ExpiresIn = accessToken.ExpiresOn.UtcDateTime,
+                AccessToken = accessToken.Token
+            };
+        }
+
+        private void AddToCache(string key, AccessTokenItem accessTokenItem)
```

Comment:
Kindly change name of method from **AddToCache** To **AddAuthTokenToCache** for more clarity.

Reply:
Yes did the changes accordingly